src/components/ElementBar/DuplicateActiveElementButton/DuplicateActiveElementButton.tsx

```diff
@@ -15,70 +15,79 @@
  */
 
 import ContentCopyIcon from '@mui/icons-material/ContentCopy';
-import log from 'loglevel';
 import { useCallback } from 'react';
 import { useTranslation } from 'react-i18next';
 import {
-  calculateBoundingRectForPoints,
   Element,
-  useActiveElement,
+  useActiveElements,
   useWhiteboardSlideInstance,
+  WhiteboardSlideInstance,
 } from '../../../state';
+import { calculateBoundingRectForElements } from '../../../state/crdt/documents/elements';
+import { BoundingRect } from '../../../state/crdt/documents/point';
 import { ToolbarButton } from '../../common/Toolbar';
 import { gridCellSize, whiteboardWidth } from '../../Whiteboard';
 
 export function duplicate(
   element: Element,
   gridCellSize: number,
-): Element | undefined {
-  const activeElementX = element.position.x;
+  boundingRect: BoundingRect,
+): Element {
+  const elementPositionX = element.position.x;
+  const isDuplicationExceedingWhiteboard =
+    boundingRect.offsetX +
+      boundingRect.width +
+      gridCellSize +
+      boundingRect.width >
+    whiteboardWidth;
 
-  const calcX = (width: number) => {
-    const newX = activeElementX + width + gridCellSize;
-    return Math.min(newX, whiteboardWidth - width);
-  };
+  let x = 0;
 
-  if (element.type === 'shape') {
-    const width = element.width;
+  if (isDuplicationExceedingWhiteboard) {
+    const boundingRectOffsetX = whiteboardWidth - boundingRect.width;
+    const offsetX = boundingRectOffsetX - boundingRect.offsetX;
 
-    return {
-      ...element,
-      position: { x: calcX(width), y: element.position.y },
-    };
-  }
-  if (element.type === 'path') {
-    const { width } = calculateBoundingRectForPoints(element.points);
+    x = elementPositionX + offsetX;
+  } else {
+    const offsetX = elementPositionX - boundingRect.offsetX;
 
-    return {
-      ...element,
-      position: { x: calcX(width), y: element.position.y },
-    };
+    x = boundingRect.offsetX + boundingRect.width + gridCellSize + offsetX;
   }
-  return undefined;
+
+  return {
+    ...element,
+    position: { x, y: element.position.y },
+  };
+}
+
+export function sort(
+  slideInstance: WhiteboardSlideInstance,
+  activeElementIds: string[],
+): string[] {
```

Comment:
Would be nice to have this method outside of duplicate button, because we could use it for other multiselect operations were we need to respect the elements order (example move all elements to front/back)

Reply:
I did consider it and felt like it could be extracted when it is needed.

Anyway, where would you prefer this to be placed?

Reply:
it could be where WhiteboardSlideInstance type is located (file src/state/types.ts already some util functions there) for example